(PUP-9217) Revert the reverts, add tests, fix issue that caused the revert

ekinanp opened the pull request.

ekinanp commented:
I've verified that `should_only_fail_associated_resources_on_file_read_error.rb` does fail without the PUP-9217 changes, and I've also verified that the backwards compatibility tests pass with and without the PUP-9217 changes (so the latter should not need to be reverted in case PUP-9217 needs to be reverted again).

ekinanp commented:
jenkins please test this on centos6-64a,centos7-64a,redhat5-64a,redhat7-64a,aix71-POWERfa,solaris10-64a,solaris11-64a,ubuntu1604-32a,ubuntu1804-64a,osx1012-64a,debian8-64a,debian9-64a,fedora27-64a,sles12-64a,sles15-64a

ekinanp commented on lib/puppet/util/filetype.rb:
```diff
@@ -191,13 +191,21 @@ def path=(user)
         return ""
       when /are not allowed to/
-        raise FileReadError, _("User %{path} not authorized to use cron") % { path: @path }, detail.backtrace
+        Puppet.err _("The %{path} user is not authorized to use cron. Their crontab file is treated as empty in case Puppet modifies e.g. cron.deny or cron.allow in the middle of the run.") % { path: @path }
```
Any thoughts on the error message here? Was trying to keep it consistent with what the original crontab provider did.

ekinanp replied:
Could also make it a debug message to keep it consistent with the unknown user case.

ekinanp replied:
Talked to @branan -- this should be debug

ekinanp commented on lib/puppet/util/filetype.rb:
```diff
@@ -191,13 +191,21 @@ def path=(user)
     # Read a specific @path's cron tab.
     def read
+      unless Puppet::Util.uid(@path)
+        Puppet.debug _("The %{path} user does not exist. Treating their crontab file as empty in case Puppet creates them in the middle of the run.") % { path: @path }
```
How about the debug message here? Originally an error message was printed, but this isn't really an error case as the workflow of "Create new user => create their crontab" is an acceptable scenario.

ekinanp replied:
Talked to @branan -- debug is fine here.

ekinanp commented:
Looks like only the `should_fail_to_write_a_nonexistent_users_crontab.rb` test failed, and that's on the platforms that rely on the crontab filetype. Pushed up a commit with the fix and the reason behind the failure -- TLDR is that it is unrelated to the PUP-9217 changes and is a bug in the crontab filetype.

ekinanp commented:
jenkins please test this on centos6-64a,centos7-64a,redhat5-64a,redhat7-64a,aix71-POWERfa,solaris10-64a,solaris11-64a,ubuntu1604-32a,ubuntu1804-64a,osx1012-64a,debian8-64a,debian9-64a,fedora27-64a,sles12-64a,sles15-64a

puppetcla commented:
CLA signed by all contributors.

ekinanp commented:
PR tests pass, this should be good to go now.

caseywilliams commented on lib/puppet/util/filetype.rb:
```diff
@@ -283,13 +307,21 @@ def write(text)
-      when /You are not authorized to use the cron command/
-        raise FileReadError, _("User %{path} not authorized to use cron") % { path: @path }, detail.backtrace
+      when /not.*authorized.*cron/
+        Puppet.err _("The %{path} user is not authorized to use cron. Their crontab file is treated as empty in case Puppet modifies e.g. cron.deny or cron.allow in the middle of the run.") % { path: @path }
```
Docs guidelines suggest avoiding latin abbreviations like `e.g.` and using abbreviated English phrasing like "for example" instead.

ekinanp replied:
Thanks, I reworded the message.

ekinanp commented:
I updated the PR to use a `debug` message instead of an `error` message for an unauthorized user in the crontab filetypes. Also changed the wording of the message to be more consistent with the docs guidelines.

ekinanp commented on acceptance/tests/resource/cron/should_only_fail_associated_resources_on_file_read_error.rb:
```diff
@@ -0,0 +1,117 @@
+      create_remote_file(agent, apply_crontab_overwrite_manifest, script)
+      on(agent, "chmod a+x #{apply_crontab_overwrite_manifest}")
+
+      puppet_result = on(agent, "bash #{apply_crontab_overwrite_manifest}", environment: { PATH: "#{stub_crontab_bin_dir}:#{path_env_var}" })
```
We do need to use a script, because the PATH variable does not persist in the `&&`. Specifically, what happens is:
```
env PATH="/tmp/stub_crontab_bin_dir.tEaOlQ:$PATH:/opt/csw/bin:/usr/sbin:/usr/bin:/opt/puppetlabs/bin" cd /tmp/stub_crontab_bin_dir.tEaOlQ && puppet apply /tmp/crontab_overwrite_manifest.r8aqmQ
```

is what's run, but only the `cd` gets the PATH variable. Our puppet apply command doesn't.
